### PR TITLE
Implement auth

### PR DESCRIPTION
**Pull request by minglis**
First pass at implementing API authentication using new JWT tokens …
- NOTE - this does not manage secrets. There is only one URL and there is no functionality implemented
- prior to rolling out we need to store secrets properly

Uses the JWT libraries in [https://github.com/alphagov/notifications-python-client](https://github.com/alphagov/notifications-python-client)

- Tokens are checked on every request and will be rejected if token is invalid as per the rules in the python clients.

### CODE REVIEW

File: app/main/authentication/auth.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    auth_scheme = auth_header[:7]
+
+    if auth_scheme != 'Bearer ':
+        return authentication_response('Unauthorized, authentication bearer scheme must be used', 401)
+
+    try:
+        auth_token = auth_header[7:]
+        api_client = fetch_client(get_token_issuer(auth_token))
+
+        if api_client is None:
```
Comment (robyoung) [inline app/main/authentication/auth.py]: This cannot be true, `fetch_client` always returns a dictionary.

File: app/main/authentication/auth.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+from flask import request, jsonify, _request_ctx_stack
+from client.jwt import decode_jwt_token, get_token_issuer
```
Comment (robyoung) [inline app/main/authentication/auth.py]: Generic names for top level packages can become a pain if you have a clash.

File: app/main/__init__.py [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
+AUTHORIZATION_SCHEME = 'Bearer'
+WINDOW = 1
 
 main = Blueprint('main', __name__)
 
 
+main.before_request(requires_auth)
+
+
 from .views import notifications, index
```
Comment (servingUpAces) [inline app/main/__init__.py]: Can we use absolute imports, please?

File: app/main/errors.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+from flask import jsonify
+
+from . import main
```
Comment (servingUpAces) [inline app/main/errors.py]: Use absolute path for the import

**Comment by servingUpAces** [comment]
Do we want to build this on travis?

**Pull request merged**